lm_eval/base.py

```diff
@@ -451,34 +456,121 @@ def higher_is_better(self):
 
         example = self.doc_to_text(doc)
         return description + labeled_examples + example
+    
+    def read_memory(self, retrieval_path):
+        # Load pre-computed nearest neighbor index for each valid dataset
+        if self._memory is None:
+            print(f'Build memory from {retrieval_path}')
+            with open(retrieval_path, 'rb') as f:
+                self._memory = pickle.load(f)
+        return self._memory
+                
+    def sample_retreival_idx(self, doc, num_fewshot, retrieval_method, retrieval_path):   
+        
+        def _compute_prob_sum(retrieval_idx, retrieval_method, sim_key, max_sim):
+            sum_sim = 0.0
+            
+            for r in retrieval_idx:
+                if retrieval_method == 'dense':
+                    norm_sim = 1 / (r[sim_key]/max_sim)
+                    sum_sim += norm_sim
+                elif retrieval_method == 'bm25':
+                    # norm_sim = math.sqrt(r[sim_key]/max_sim)
+                    norm_sim = r[sim_key]/max_sim                
+                    sum_sim += norm_sim  
+            return sum_sim
+        
+        memory = self.read_memory(retrieval_path)
+        doc_idx = doc['doc_idx']        
+        retrieval_idx = memory[doc_idx]['retrieval_idx'][:4*num_fewshot]
+        # print(retrieval_idx)
+        # print('@@@ Length of retrieval_idx: {}'.format(len(retrieval_idx)))
+        
+        if retrieval_method == 'dense':
+            sim_key = 'l2_distance'
+        elif retrieval_method == 'bm25':
+            sim_key = 'score'
+        
+        # Normalize prob s.t the sum equals to 1.0
+        max_sim = retrieval_idx[0][sim_key]
+        # print('@@@ max_sim: {}'.format(max_sim))
+        
+        sum_sim = _compute_prob_sum(retrieval_idx, retrieval_method, sim_key, max_sim)
+        for r in retrieval_idx:
+            if retrieval_method == 'dense':
+                norm_sim = 1 / (r[sim_key]/max_sim)
+                r[sim_key] = norm_sim / sum_sim
+            elif retrieval_method == 'bm25':
+                # norm_sim = math.sqrt(r[sim_key]/max_sim)
+                norm_sim = r[sim_key]/max_sim
+                r[sim_key] = norm_sim / sum_sim        
+                
+        list_of_index = [r['index'] for r in retrieval_idx]
+        list_of_prob = [r[sim_key] for r in retrieval_idx]
+        
+        # print('>>> doc_idx: ', doc_idx)
+        # print('>>> list_of_index: ',list_of_index)
+        # print('>>> list_of_prob: ',list_of_prob)        
+        # print(' ')
+        
+        try:
+            top_k_idx = np.random.choice(list_of_index, num_fewshot, replace=False, p=list_of_prob)
+        except:
+            # This arises when there are suffcient 0.0 values
+            # https://stackoverflow.com/questions/57890844/python-numpy-random-choice-valueerror-fewer-non-zero-entries-in-p-than-size
+            top_k_idx = np.random.choice(list_of_index, num_fewshot, replace=True, p=list_of_prob)
```

Comment:
```suggestion
      replace = False if len(list_of_index) >= num_fewshot else True
      top_k_idx = np.random.choice(list_of_index, num_fewshot, replace=replace, p=list_of_prob)
```

요러한 로직을 원하시는게 맞는지는 모르겠네요.
뭔가 얘외처리보다 좀더 명시적으로 상황에 대한 설명이 될 수 있으면 좋을것 같아서 수정해보았습니닷